File: skiko/src/jvmMain/java/org/jetbrains/awthrl/DriverApi/Engine.java
package org.jetbrains.awthrl.DriverApi;

import org.jetbrains.awthrl.Common.OSType;
import org.jetbrains.awthrl.Components.Drawable;

public class Engine {
    private Engine() {
    }

    private static Engine instance = null;

    private GraphicsApi api = GraphicsApi.UNKNOWN;
    public GraphicsApi currentApi() {
        return api;
    }

    public static Engine get() {
        if (instance == null) {
            System.loadLibrary("drawcanvas");
            instance = new Engine();
            instance.initSuitableGraphicsApi(OSType.getCurrent());
        }
        return instance;
    }

    /**
     * GAG - At the moment always initializes OpenGL.
     * @param osType The current OS type of the device.
     */
    private void initSuitableGraphicsApi(OSType osType) {
        if (osType == OSType.MAC_OS) {
            OpenGLApi.get(); // for Metal
            instance.api = GraphicsApi.OPENGL;
        } else if (osType == OSType.LINUX) {
            OpenGLApi.get(); // for Vulkan and OpenGL
            instance.api = GraphicsApi.OPENGL;
        } else if (osType == OSType.WINDOWS) {
            OpenGLApi.get(); // for Vulkan and OpenGL
            instance.api = GraphicsApi.OPENGL;
        } else {
            OpenGLApi.get(); // default
            instance.api = GraphicsApi.OPENGL;
        }
    }

    public void render(Drawable drawable) {
        drawable.updateLayer();
        drawable.redrawLayer();
    }
}

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Engine.kt
package org.jetbrains.skiko

class Engine private constructor() {
    private var api = GraphicsApi.UNKNOWN
    fun currentApi(): GraphicsApi {
        return api
    }

    /**
     * GAG - At the moment always initializes OpenGL.
     * @param osType The current OS type of the device.
     */
    private fun initSuitableGraphicsApi(osType: OSType) {
        when {
            osType === OSType.MAC_OS -> {
                OpenGLApi.get() // for Metal
                instance.api = GraphicsApi.OPENGL
            }
            osType === OSType.LINUX -> {
                OpenGLApi.get() // for Vulkan and OpenGL
                instance.api = GraphicsApi.OPENGL
            }
            osType === OSType.WINDOWS -> {
                OpenGLApi.get() // for Vulkan and OpenGL
                instance.api = GraphicsApi.OPENGL
            }
            else -> {
                OpenGLApi.get() // default
                instance.api = GraphicsApi.OPENGL
            }
        }
    }

    fun render(drawable: Drawable) {
        drawable.updateLayer()
        drawable.redrawLayer()
    }

    companion object {
        private lateinit var instance: Engine
        fun get(): Engine {
            if (!this::instance.isInitialized) {
                instance = Engine()
                instance.initSuitableGraphicsApi(OSType.getCurrent())
            }
            return instance
        }
    }
}
